swarm/src/handler/either.rs

```diff
@@ -90,6 +92,207 @@ impl<L, R> IntoEitherHandler<L, R> {
     }
 }
 
+impl<'a, RIP, LIP, LOP, ROP, LIOP, RIOP, LOOP, ROOP>
+    ConnectionEvent<
+        'a,
+        EitherUpgrade<SendWrapper<LIP>, SendWrapper<RIP>>,
+        EitherUpgrade<SendWrapper<LOP>, SendWrapper<ROP>>,
+        Either<LIOP, RIOP>,
+        Either<LOOP, ROOP>,
+    >
+where
+    RIP: InboundUpgradeSend,
+    LIP: InboundUpgradeSend,
+    LOP: OutboundUpgradeSend,
+    ROP: OutboundUpgradeSend,
+{
+    /// Transposes a [`ConnectionEvent`] with Either's on it's inner fields to an [`Either`] of connection events.
+    /// Returns Err in case of the [`ConnectionEvent`] being [`AddressChange`] as it isn't transposable.
+    pub fn transpose(
+        self,
+    ) -> Result<
+        Either<
+            ConnectionEvent<'a, LIP, LOP, LIOP, LOOP>,
+            ConnectionEvent<'a, RIP, ROP, RIOP, ROOP>,
+        >,
+        AddressChange<'a>,
+    > {
+        match self {
+            ConnectionEvent::FullyNegotiatedInbound(FullyNegotiatedInbound {
+                protocol: EitherOutput::First(protocol),
+                info: Either::Left(info),
+            }) => Ok(Either::Left(ConnectionEvent::FullyNegotiatedInbound(
+                FullyNegotiatedInbound { protocol, info },
+            ))),
```

Comment:
```suggestion
            ConnectionEvent::FullyNegotiatedInbound(inner) => Ok(inner.transpose()),
```

Does this work? Or do we have to call `.map` on it again or something?

It would be cool if we can express this `transpose` function somehow through the smaller ones.

Reply:
It doesn't because `Select`'s implementation uses `SelectUpgrade` whereas `Either` uses `EitherUpgrade`, do you have any idea how can we combine them in `transpose`? :P

Reply:
I don't think we can combine them for select and either because those are fundamentally different.

Reply:
Also, one of them is using `EitherOutput` which we need because `Either` doesn't implement `AsyncRead` + `AsyncWrite`.

See https://github.com/libp2p/rust-libp2p/issues/2650.